// hyperrequest/lib/main.js
(function() {
  var HyperRequest, StringDecoder, hrquest, http, hyperquest, isString, querystring,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  querystring = require("querystring");

  StringDecoder = require('string_decoder').StringDecoder;

  http = require("http");

  hyperquest = require("hyperquest");

  http.globalAgent.maxSockets = 30;

  isString = function(vr) {
    return typeof vr === 'string' || vr instanceof String;
  };

  hrquest = new (HyperRequest = (function(superClass) {
    extend(HyperRequest, superClass);

    function HyperRequest() {
      this.ERRORS = bind(this.ERRORS, this);
      this._responseHandler = bind(this._responseHandler, this);
      this.prepareOpts = bind(this.prepareOpts, this);
      this.prepareBody = bind(this.prepareBody, this);
      this.request = bind(this.request, this);
      this.defaults = bind(this.defaults, this);
      return HyperRequest.__super__.constructor.apply(this, arguments);
    }

    HyperRequest.prototype.defaults = function() {
      return this.extend(HyperRequest.__super__.defaults.apply(this, arguments), {
        method: "GET",
        payloadMethods: ["POST", "PUT", "PATCH", "DELETE"]
      });
    };

    HyperRequest.prototype.request = function(opt, cb) {
      this.prepareBody(opt, (function(_this) {
        return function(err, _body, _forcedHeaders) {
          var _err, _opts, _path, error, ref, request;
          if (err) {
            cb(err);
            return;
          }
          try {
            _path = _this.preparePath(opt);
            _opts = _this.prepareOpts(opt, _body, _forcedHeaders);
          } catch (error) {
            _err = error;
            cb(_err);
            return;
          }
          _this.debug("run request", _path, _opts, _body, _body != null ? _body.toString() : void 0);
          request = hyperquest(_path, _opts);
          _this._responseHandler(cb, request, _path, _opts);
          if (_body != null) {
            request.write(_body);
          }
          if (ref = _opts.method, indexOf.call(_this.config.payloadMethods, ref) >= 0) {
            request.end();
          }
        };
      })(this));
    };

    HyperRequest.prototype.prepareBody = function(opt, cb) {
      var _body, _e, _fHeaders, _sjson, error;
      _body = null;
      _fHeaders = {};
      if (opt.json != null) {
        if (isString(opt.json)) {
          _body = new Buffer(opt.json);
        } else {
          try {
            _sjson = JSON.stringify(opt.json);
          } catch (error) {
            _e = error;
            this._handleError(cb, "json-stringify");
            return;
          }
          _body = new Buffer(_sjson);
        }
        _fHeaders["Content-type"] = "application/json";
      } else if (opt.body != null) {
        if (Buffer.isBuffer(opt.body)) {
          _body = opt.body;
        } else if (isString(opt.body)) {
          _body = new Buffer(opt.body);
        } else {
          this.warning("used body unlike Buffer or String!", opt.body);
          _body = opt.body;
        }
      }
      if (_body != null ? _body.length : void 0) {
        _fHeaders["Content-length"] = _body.length;
      }
      cb(null, _body, _fHeaders);
    };

    HyperRequest.prototype.preparePath = function(opt) {
      var _qs, _url;
      _url = opt.url || opt.uri || opt.path;
      _qs = "";
      if (opt.qs != null) {
        if (isString(opt.qs)) {
          _qs = opt.qs;
        } else {
          _qs = querystring.stringify(opt.qs);
        }
      }
      if (_qs != null ? _qs.length : void 0) {
        if (_url.indexOf("?") < 0) {
          return _url + "?" + _qs;
        } else {
          return _url + "&" + _qs;
        }
      } else {
        return _url;
      }
    };

    HyperRequest.prototype.prepareOpts = function(opt, _body, _forcedHeaders) {
      var _enc, _method, _pw, _user, ref;
      _method = ((ref = opt.method) != null ? ref.toUpperCase() : void 0) || this.config.method;
      if ((_body != null) && indexOf.call(this.config.payloadMethods, _method) < 0) {
        this._handleError(null, "invalid-method");
        return;
      }
      if ((opt != null ? opt.auth : void 0) != null) {
        _user = opt.auth.user || opt.auth.username;
        _pw = opt.auth.pass || opt.auth.password || "";
        if (_user != null) {
          _enc = new Buffer(_user + ":" + _pw, "utf8").toString("base64");
          _forcedHeaders["Authorization"] = "Basic " + _enc;
        }
      }
      return {
        method: _method,
        headers: this.extend({}, opt.headers || {}, _forcedHeaders)
      };
    };

    HyperRequest.prototype._responseHandler = function(cb, request, _path, _opts) {
      var _body, _err, decoder, res;
      _err = "";
      _body = "";
      res = null;
      decoder = new StringDecoder("utf8");
      request.on("response", (function(_this) {
        return function(response) {
          res = response;
          _this.debug("request response", _path, _opts, _body, _body != null ? _body.toString() : void 0);
        };
      })(this));
      request.on("data", function(chunk) {
        _body += decoder.write(chunk);
      });
      request.on("error", (function(_this) {
        return function(chunk) {
          var ref;
          _this.warning("request error", chunk.toString());
          if ((ref = chunk != null ? chunk.code : void 0) === "ECONNREFUSED" || ref === "EHOSTUNREACH") {
            cb(chunk);
            return;
          }
          _err += chunk;
        };
      })(this));
      request.on("end", (function(_this) {
        return function() {
          var ref, ref1, ref2;
          _this.debug("request end", _body, _err);
          if (_err) {
            cb(_err, res);
            return;
          }
          _this.debug("request result", res.statusCode, res.headers, _body);
          if (!(_body != null ? _body.length : void 0)) {
            res.body = null;
          } else if ((_opts != null ? (ref = _opts["Content-type"]) != null ? ref.toLowerCase() : void 0 : void 0) === "application/json" || ((ref1 = res.headers) != null ? (ref2 = ref1["content-type"]) != null ? typeof ref2.indexOf === "function" ? ref2.indexOf("application/json") : void 0 : void 0 : void 0) >= 0) {
            try {
              res.body = JSON.parse(_body);
            } catch (undefined) {}
          } else if (_body != null ? _body.length : void 0) {
            res.body = _body;
          }
          cb(null, res);
        };
      })(this));
      request.on("close", (function(_this) {
        return function() {
          _this.debug("request close", _body, _err);
        };
      })(this));
    };

    HyperRequest.prototype.ERRORS = function() {
      return this.extend(HyperRequest.__super__.ERRORS.apply(this, arguments), {
        "json-stringify": [500, "Cannot stringify the given `json` data."],
        "invalid-method": [500, "If you pass body data via `body` or `json` you have to use a method of `POST, `PUT` or `PATCH`"]
      });
    };

    return HyperRequest;

  })(require("mpbasic")()))();

  module.exports = hrquest.request;

}).call(this);
